File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/655-combinations.java
package main.java.Medium;

import java.util.ArrayList;
import java.util.List;

/**https://leetcode.com/problems/combinations/ */
//Given two integers n and k, return all possible combinations of k numbers chosen from the range [1, n].
//You may return the answer in any order.
// 
//Example 1:
//Input: n = 4, k = 2
//Output: [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
//Explanation: There are 4 choose 2 = 6 total combinations.
//Note that combinations are unordered, i.e., [1,2] and [2,1] are considered to be the same combination.
//Example 2:
//Input: n = 1, k = 1
//Output: [[1]]
//Explanation: There is 1 choose 1 = 1 total combination.
// 
//Constraints:
//	1 <= n <= 20
//	1 <= k <= n
class Solution452 {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(n, k, 1, new ArrayList<>(), res);
        return res;
    }
    public void dfs(int n, int k, int start, List<Integer> path, List<List<Integer>> res) {
        if (path.size() == k) {
            res.add(new ArrayList<>(path));
        }
        for (int i = start; i <= n; i++) {
            path.add(i);
            dfs(n, k, i + 1, path, res);
            path.remove(path.size() - 1);
        }
    }
}

File: ScanProject/PyAnalysis/SonarPy/Medium/655-combinations.py
#https://leetcode.com/problems/combinations/
#Given two integers n and k, return all possible combinations of k numbers chosen from the range [1, n].
#You may return the answer in any order.


class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """ 
        def dfs(nums, k, start, path, res):
            if len(path) == k:
                res.append(path[:])
                return
            for i in range(start, len(nums)):
                path.append(nums[i])
                dfs(nums, k, i + 1, path, res)
                path.pop()
        res = []
        dfs(range(1, n + 1), k, 0, [], res)
        return res

    